pallets/auctions/src/lib.rs

```diff
@@ -0,0 +1,852 @@
+					auction_object.create(sender, &auction)?;
+				}
+				Auction::TopUp(auction_object) => {
+					auction_object.create(sender, &auction)?;
+				}
+				Auction::Candle(auction_object) => {
+					auction_object.create(sender, &auction)?;
+				}
+			}
+
+			Ok(())
+		}
+
+		///
+		/// Updates an existing auction which has not yet started
+		///
+		/// - calls the update() implementation on the given Auction type
+		///
+		#[pallet::weight(
+			<T as Config>::WeightInfo::update_english()
+				.max(<T as Config>::WeightInfo::update_topup())
+				.max(<T as Config>::WeightInfo::update_candle())
+		)]
+		#[transactional]
+		pub fn update(origin: OriginFor<T>, id: T::AuctionId, updated_auction: Auction<T>) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			match updated_auction {
+				Auction::English(auction_object) => {
+					auction_object.update(sender, id)?;
+				}
+				Auction::TopUp(auction_object) => {
+					auction_object.update(sender, id)?;
+				}
+				Auction::Candle(auction_object) => {
+					auction_object.update(sender, id)?;
+				}
+			}
+			Ok(())
+		}
+
+		///
+		/// Destroys an existing auction which has not yet started
+		///
+		/// - validates write action
+		/// - unfreezes NFT
+		/// - calls destroy helper function
+		///
+		#[pallet::weight(
+			<T as Config>::WeightInfo::destroy_english()
+				.max(<T as Config>::WeightInfo::destroy_topup())
+				.max(<T as Config>::WeightInfo::destroy_candle())
+		)]
+		#[transactional]
+		pub fn destroy(origin: OriginFor<T>, id: T::AuctionId) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let auction = <Auctions<T>>::get(id).ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+			match &auction {
+				Auction::English(auction_object) => {
+					Self::validate_update(sender, &auction_object.common_data)?;
+					Self::unfreeze_nft(&auction_object.common_data)?;
+					Self::handle_destroy(id)?;
+				}
+				Auction::TopUp(auction_object) => {
+					Self::validate_update(sender, &auction_object.common_data)?;
+					Self::unfreeze_nft(&auction_object.common_data)?;
+					Self::handle_destroy(id)?;
+				}
+				Auction::Candle(auction_object) => {
+					Self::validate_update(sender, &auction_object.common_data)?;
+					Self::unfreeze_nft(&auction_object.common_data)?;
+					Self::handle_destroy(id)?;
+				}
+			}
+
+			Ok(())
+		}
+
+		///
+		/// Places a bid on a running auction
+		///
+		/// - validates bid
+		/// - calls the bid() implementation on the given Auction type
+		/// - deposits BidPlaced event
+		///
+		#[pallet::weight(
+			<T as Config>::WeightInfo::bid_english()
+				.max(<T as Config>::WeightInfo::bid_topup())
+				.max(<T as Config>::WeightInfo::bid_candle())
+		)]
+		#[transactional]
+		pub fn bid(origin: OriginFor<T>, auction_id: T::AuctionId, amount: BalanceOf<T>) -> DispatchResult {
+			let bidder = ensure_signed(origin)?;
+			let bid = Bid {
+				amount,
+				block_number: frame_system::Pallet::<T>::block_number(),
```

Comment:
i would consider using relay chain block numbers for more predictable conversion to time

Reply:
+100